content/advent-2018/split-data-with-cdc.md

```diff
@@ -0,0 +1,232 @@
+
+How can we use that to find chunk boundaries? Restic uses a window of 64 bytes, so while reading a file, it computes the fingerprint of all such windows: byte 0 to byte 63, byte 1 to byte 64 and so on. It then checks every fingerprint if the least significant bits are all zero. If this is the case, then we found a chunk boundary! For random input bytes, the calculated fingerprints are roughly random as well, so by varying the number of bits that are checked we can (roughly) configure how large the resulting chunks will be. Since the cut points for individual chunks only depend on the 64 bytes preceding them, this method of splitting data is called "Content-Defined Chunking" (CDC).
+
+Early on I recognized that the algorithm and implementation has the potential to be very helpful to other projects as well, so I published it as a [separate package](https://github.com/restic/chunker) for other people to import ([API on godoc.org](https://godoc.org/github.com/restic/chunker)) with a very liberal BSD 2-clause license.
+
+Enough theory, let's dive into the code! 
+
+Splitting data
+==============
+
+The type which does all the hard work of computing the fingerprints over the sliding window is contained in the `Chunker` type. The Rabin fingerprint needs a so-called "irreducible polynomial" to work, which is encoded as an `uint64`. The `chunker` package exports the function `RandomPolynomial()` which creates a new polynomial for you to use. **Important:** Different polynomials will result in different fingerprints and therefore different boundaries for the chunks, so if you depend on your program computing the same boundaries, you'll need to pass in the same polynomial.
+
+Our first program will generate a random polynomial and create a new `Chunker` which reads data from `os.Stdin`:
+
+```go
+p, err := chunker.RandomPolynomial()
+if err != nil {
+	panic(err)
+}
+
+fmt.Printf("using polynomial %v for splitting data\n", p)
+
+chunker := chunker.New(os.Stdin, p)
+```
+
+By default, it tries to make chunks of one megabyte on average (you can control that with the method `SetAverageBits()`). Then we can repeatedly call the `Next()` function on the `Chunker` and pass in a byte slice buffer for the data read from the reader passed to `New()`, it'll return a `Chunk` type and an error:
```

Comment:
> "By default, it tries to make chunks of one megabyte on average <..>

Do you think this is important for this paragraph? I feel that it can be omitted here, leaving a reader focused on what is happening in the example.

---

Can I suggest rephrasing the next sentence to something like

"The `Next()` *method* on the `Chunker` reads data out the reader into a byte slice buffer. The method returns a next `Chunk` and an error. We call the method untill it returns `io.EOF` error, to read all the data:"

Reply:
Both good points, thanks for the ideas! And you're right, it's a *method* ;)